### ISSUE TITLE
[question] Conan 2 lockfiles: How-to lock pkgid and prev?

### ISSUE THREAD

**Issue by SzBosch** [conan-io/docs#3869]
Dear Conan team,

earlier with Conan 1 a lockfile contained the full reference incl. pkgid and prev.

With Conan 2.0.x this information is not anymore in the lockfile by default. But you provided a parameter "--lockfile-packages" which brought back this information.

Now with newer Conan versions (tried 2.4.1 and 2.6.0) this parameter is gone from the documentation, but can be still given on the command line, but giving an (non-breaking) error, e.g.:
```
conan lock create --requires="mypkg/1.0.0" --lockfile-packages
ERROR: The --lockfile-packages arg is private and shouldn't be used
...
```

What is the correct way to get a full lockfile incl. pkgid and prev?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by d3matt** [comment]
@memsharded (sorry to resurrect a months old issue) if multiple binary packages are really to be avoided... Could `conan upload` default to not upload binary packages that match something already uploaded? I'm nervous that there's a bit of a hole here for people who have not setup CI for uploading conan packages but still want to have dependency level reproducibility of a build from a conan lock file. With the v1 lock files, I had a guarantee that I would get the exact same binaries every time for my dependencies.

**Comment by memsharded** [comment]
Hi @d3matt and all,

Yes, sorry this has fallen through the cracks, sometimes it happens when things are too busy.
Yes, we have been wanting to add something to improve the "protection" to avoid this situation, but we couldn't prioritize it yet. Thanks for your ping, I hope we can have a look to that soon.

**Comment by adnanali21** [comment]
Agree with @hpe-ykoehler  and @HekyM on this topic. Lets not have this dropped out as it is a guarantee for the reproducible builds which is basically one of the most (if not the most) important thought behind having a lockfile. IMO if we can have one leakage where a workflow happens without me controlling it even with lockfile, the whole reason behind having a lockfile is useless.

**Comment by uboot** [comment]
Short comment from my side: we have established a workflow where we use a Conan recipe to archive data from a network share into a Conan package (no build, only packaging). If the content of the network share changes in a significant way we build a new package. The only way to distinguish packages built at different times is via their PREV (the RREV does not change). If we would want to lock such a package the PREV would be needed in the lockfile.

Obviously, this scenario can be considered a mis-use of Conan and probably is. I just wanted to give some more context to this topic.

**Comment by p-groarke** [comment]
@uboot The concept of multiple PREVs has been removed in conan 2 (as it should be). Having different PREVs for the same input should be considered a bug in your recipe. You might want to simply bump the version number every time you release a new package :)

@memsharded We also need `package_id` in our lockfiles (not PREVs, PREVs are evil and removing them from conan 2 is one of the best decisions made). We need them for a really weird reason however, so here's why.

We need the `package_ids` in our lockfiles to trigger merge conflicts. Yes I know that's unorthodox, might be an "abuse" of lockfiles, but it's actually a major plus for us. You may have situations where multiple concurrent PRs are open requiring rebuilds, and whoever merges first, we want the lockfile to report merge conflicts to the other PRs (forcing rebuilds).

IMO, if you try to use a lockfile with a config and profile that doesn't generate an identical  `package_id`, it should be considered an error and fail. Since as you say, it is now (thankfully 🥳 ) deterministic. So the `package_id` is really just "documentation" in the lockfile.

I know how easy it would be to simply add this option to `lock create`, so please consider it for package_ids :] PREVs should *not* be encoded in the lockfile with this option (could be a seperate option or altogether unavailable), since if you need to lock PREVs you have a bug. aka PREVs are evil. ty!

**Comment by uboot** [comment]
Hm, the Conan 2 I have is actually pretty good at handling PREVs  ;-).

@p-groarke  I agree that multiple PREVs should be avoided in an efficient CI/CD setup. I personally would not call them evil, though. Do you know any situations where having multiple PREVs leads to serious consequences (breaking CI, loss of data, shipping wrong/not tested artifacts etc.)? If so this might only be an even stronger reason to lock the PREV.

**Comment by p-groarke** [comment]
@uboot We want the same thing, I am just advocating to split lock options into `--lock-pids` + `--lock-prevs` options since locking PREVs is giving you ammunition to shoot yourself in the foot ;)

Worded differently, why would you allow locking PREVs if there is only 1 PREV for a given recipe input?

In conan, your reference (package, ver, channel, user) + rrev + package_id is your input. PREV is the output. For a given input, the output should always be the same. Your recipe *must* be deterministic, or else your aren't modelling the entire input state in your recipe. 2 + 2 = 4, not just sometimes, always.

And yes, we've had PREV-pocalypse before; 70+ devs unable to build their software for days while we work around the clock to fix the PREV horror show. Conan 2 does the right thing now XD

**Comment by uboot** [comment]
@p-groarke Yes, I agree. Having two lock options is a more flexible approach.
